lightning/src/ln/channelmanager.rs

```diff
@@ -1857,98 +1857,150 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref> ChannelMan
+			//
+			// If we find an HTLC which we would need to claim but for which we do not have a
+			// channel, we will fail all parts of the MPP payment. While we could maybe wait and
+			// see if the sender retries the already-failed path(s), it should be a pretty rare
+			// case what we got all the HTLCs and then a channel closed while we were waiting for
+			// the user to provide the preimage, so worrying too much about the optimal handling
+			// isn't worth it.
+
+			let (is_mpp, mut valid_mpp) = if let &Some(ref data) = &sources[0].payment_data {
 				assert!(payment_secret.is_some());
-				data.total_msat == expected_amount
+				(true, data.total_msat == expected_amount)
 			} else {
 				assert!(payment_secret.is_none());
-				false
+				(false, false)
 			};
 
+			for htlc in sources.iter() {
+				if !is_mpp || !valid_mpp { break; }
+				if let None = channel_state.as_ref().unwrap().short_to_id.get(&htlc.prev_hop.short_channel_id) {
+					valid_mpp = false;
+				}
+			}
+
+			let mut errs = Vec::new();
 			let mut one_claimed = false;
 			for htlc in sources.drain(..) {
 				if channel_state.is_none() { channel_state = Some(self.channel_state.lock().unwrap()); }
-				if !valid_mpp_amount && (htlc.value < expected_amount || htlc.value > expected_amount * 2) {
+				if (is_mpp && !valid_mpp) || (!is_mpp && (htlc.value < expected_amount || htlc.value > expected_amount * 2)) {
 					let mut htlc_msat_data = byte_utils::be64_to_array(htlc.value).to_vec();
 					let mut height_data = byte_utils::be32_to_array(self.latest_block_height.load(Ordering::Acquire) as u32).to_vec();
 					htlc_msat_data.append(&mut height_data);
 					self.fail_htlc_backwards_internal(channel_state.take().unwrap(),
 									 HTLCSource::PreviousHopData(htlc.prev_hop), &payment_hash,
 									 HTLCFailReason::Reason { failure_code: 0x4000|15, data: htlc_msat_data });
 				} else {
-					self.claim_funds_internal(channel_state.take().unwrap(), HTLCSource::PreviousHopData(htlc.prev_hop), payment_preimage);
-					one_claimed = true;
+					match self.claim_funds_from_hop(channel_state.as_mut().unwrap(), htlc.prev_hop, payment_preimage) {
+						Err(Some(e)) => {
+							if let msgs::ErrorAction::IgnoreError = e.1.err.action {
+								// We got a temporary failure updating monitor, but will clam the
+								// HTLC when the monitor updating is restored (or on chain).
+								log_error!(self, "Temporary failure claiming HTLC, treating as success: {}", e.1.err.err);
+								one_claimed = true;
+							} else { errs.push(e); }
+						},
+						Err(None) if is_mpp => unreachable!("We already checked for channel existence, we can't fail here!"),
+						Err(None) => { /* This is reachable if the channel closed while the user was getting the preimage */ },
+						Ok(()) => one_claimed = true,
+					}
 				}
 			}
+
+			// Now that we've done the entire above loop in one lock, we can handle any errors
+			// which were generated.
+			channel_state.take();
```

Comment:
It is true because 1) for MPP, either all or no HTLCs need to fail backward, and 2) for single-path we don't care, this comment was only intended to refer to MPP :). I agree it would be nice to separate out, but I didn't see an obvious way to do it that didn't duplicate most of the calling logic for both.

Reply:
For my own education, why don't we care for single-path? Is (2) that I mentioned above accurate? If not, when could there be multiple HTLCs to claim for a single-path payment?

Reply:
In theory an on-path attacker (ie middle node in a route) could send payments onwards to every other node in the network with the same payment_hash. This causes an inordinate amount of complexity - HTLCs may be claimed to different sources (via HTLCSource) even with the same preimage, we can't just blindly reveal whether we know the preimage for such a payment, etc.

In this context, we do that via the HTLC's value (as hopefully the attacker isn't willing to actually pay us real value to deanonymize who the recipient of a payment is), though long-term this is the reason for the payment_secret to exist.